src/charm.py

```diff
@@ -2,22 +2,200 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-"""Charmed operator for the 5G NSSF service."""
+"""Charmed operator for the SD-Core NSSF service."""
 
 import logging
+from ipaddress import IPv4Address
+from subprocess import check_output
 
-from ops.charm import CharmBase
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]
+    KubernetesServicePatch,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFRequires  # type: ignore[import]
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, EventBase
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
+SBI_PORT = 29531
+CONFIG_DIR = "/free5gc/config"
+CONFIG_FILE_NAME = "nssfcfg.conf"
+CONFIG_TEMPLATE_DIR = "src/templates/"
+CONFIG_TEMPLATE_NAME = "nssfcfg.conf.j2"
+
 
 class NSSFOperatorCharm(CharmBase):
-    """Charm the service."""
+    """Main class to describe juju event handling for the SD-Core NSSF operator."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._nssf_container_name = self._nssf_service_name = "nssf"
+        self._nssf_container = self.unit.get_container(self._nssf_container_name)
+        self._nrf_requires = NRFRequires(charm=self, relation_name="fiveg_nrf")
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="sbi", port=SBI_PORT),
+            ],
+        )
+
+        self.framework.observe(self.on.nssf_pebble_ready, self._on_nssf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_nssf_pebble_ready)
+
+    def _on_nssf_pebble_ready(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for NSSF container.
+
+        Args:
+            event (EventBase): Juju event
+        """
+        if not self._nssf_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to start")
+            event.defer()
+            return
+        if not self._relation_created("fiveg_nrf"):
+            self.unit.status = BlockedStatus("Waiting for fiveg_nrf relation")
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        if not self._nssf_container.exists(path=CONFIG_DIR):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        content = self._render_config_file(
+            sbi_port=SBI_PORT,
+            nrf_url=self._nrf_requires.nrf_url,
+            nssf_url=self._nssf_hostname,
+        )
+        self._push_config_file(
+            content=content,
+        )
```

Comment:
Somewhat related to @saltiyazan comment - we should also restart a service when config changes but the pebble layer doesn't. We came up with a working solution in the [UDR charm](https://github.com/canonical/sdcore-udr-operator/blob/43e104901edd4d4c7eb2ec14001b3dc4702e7770/src/charm.py#L89)